### PR TITLE
Add autofix to font-family-name-quotes

### LINKED ISSUE #5805: Support autofix for `font-family-name-quotes` rule
### What steps are needed to reproduce the bug?

create a css file with
```css
body {
  margin: 0;
  font-family: "Roboto", sans-serif;
}
```
enable rule `'font-family-name-quotes': 'always-where-recommended'`

### What Stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "font-family-name-quotes": "always-where-recommended"
  }
}
```

### How did you run Stylelint?

`stylelint index.css --fix`

### Which version of Stylelint are you using?

`14.2.0`

### What did you expect to happen?

`"Roboto"` should be converted to `Roboto` after `--fix` flag.

### What actually happened?

`--fix` does not automatically fix the problem.

### Does the bug relate to non-standard syntax?

_No response_

### Proposal to fix the bug

I have a hunch that this was working in `stylelint@13` but I'm not sure. either way shouldn't this be an automatically fixable problem by removing the quotes on font family names that don't have spaces or misc symbols.

### PR DESCRIPTION
**Pull request by MrBrN197**
> Which issue, if any, is this issue related to?

Closes #5805

> Is there anything in the PR that needs further explanation?

- On [this line](https://github.com/MrBrN197/stylelint/blob/502e693451fe53dfe83f64ba6dd99b892e9d526c/lib/rules/font-family-name-quotes/index.js#L96) I am using the `result` object to check against the `string-quotes` rule to determine whether to replace with single or double-quotes. I'm not sure whether this is wise or not. I'm not sure what order the rules run in so if the `font-family-name-quotes` rule runs after the `string-quotes` this might cause it to replace with the wrong type of quotes and `string-quotes` will not get a chance to fix this.
- `checkFamilyName` returns an object when `context.fix` is enabled and `complain` doesn't run. modifying the `decl.value` inside `checkFamilyName` would result in the other `fontFamilyNode.sourceIndex`s' being wrong so I defer this until after the loop. I'm not sure if this is the best way to go about this. but to avoid rewriting a lot of the code this is what I decided to do. some clarification on this would be nice.

### CODE REVIEW

**Comment by ybiquitous** [comment]
@MrBrN197 Thank you for creating the pull request and sorry for the late response.

> * On [this line](https://github.com/MrBrN197/stylelint/blob/502e693451fe53dfe83f64ba6dd99b892e9d526c/lib/rules/font-family-name-quotes/index.js#L96) I am using the `result` object to check against the `string-quotes` rule to determine whether to replace with single or double-quotes. I'm not sure whether this is wise or not. I'm not sure what order the rules run in so if the `font-family-name-quotes` rule runs after the `string-quotes` this might cause it to replace with the wrong type of quotes and `string-quotes` will not get a chance to fix this.

With the current implementation, `string-quotes` is executed after `font-family-name-quotes`, that is, by the order defined `lib/rules/index.js`. See below:

https://github.com/stylelint/stylelint/blob/8a4253a56b683c32b127a612383372bb86ca99fc/lib/lintPostcssResult.js#L57-L61

https://github.com/stylelint/stylelint/blob/8a4253a56b683c32b127a612383372bb86ca99fc/lib/rules/index.js#L8

This behavior is undocumented as far as I know and may be changed in the future. However, since it will be as-is for a while, I recommend removing the code referring `config.rules['string-quotes']`.

So, it doesn't seem to be a big problem, even if `font-family-name-quotes` *always* fixes CSS with single quotes (or double quotes).


> * `checkFamilyName` returns an object when `context.fix` is enabled and `complain` doesn't run. modifying the `decl.value` inside `checkFamilyName` would result in the other `fontFamilyNode.sourceIndex`s' being wrong so I defer this until after the loop. I'm not sure if this is the best way to go about this. but to avoid rewriting a lot of the code this is what I decided to do. some clarification on this would be nice.

OK, I'll review the code more.

**Review by ybiquitous — COMMENTED** [review]
[suggest] Can you please add the following description to the `font-family-name-quotes` doc?

https://github.com/stylelint/stylelint/blob/8a4253a56b683c32b127a612383372bb86ca99fc/lib/rules/string-quotes/README.md#L28

File: lib/rules/font-family-name-quotes/__tests__/index.js [hunk 1]
```diff
@@ -390,3 +390,264 @@ testRule({
...
+	ruleName,
+	config: ['always-unless-keyword'],
+	fix: true,
+	reject: [
+		{
+			code: 'a { font-family: "Lucida Grande", "Arial", "sans-serif"; }',
+			message: messages.rejected('sans-serif'),
+			line: 1,
+			column: 45,
+			fixed: 'a { font-family: "Lucida Grande", "Arial", sans-serif; }',
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/__tests__/index.js:404]:
```suggestion
			code: 'a { font-family: "Lucida Grande", "Arial", "sans-serif"; }',
			fixed: 'a { font-family: "Lucida Grande", "Arial", sans-serif; }',
			message: messages.rejected('sans-serif'),
			line: 1,
			column: 45,
```

File: lib/rules/font-family-name-quotes/__tests__/index.js [hunk 2]
```diff
@@ -390,3 +390,264 @@ testRule({
...
+		},
+		{
+			code: 'a { font: 1em "Lucida Grande", "Arial", "sans-serif"; }',
+			message: messages.rejected('sans-serif'),
+			line: 1,
+			column: 42,
+			fixed: 'a { font: 1em "Lucida Grande", "Arial", sans-serif; }',
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/__tests__/index.js:411]:
```suggestion
			code: 'a { font: 1em "Lucida Grande", "Arial", "sans-serif"; }',
			fixed: 'a { font: 1em "Lucida Grande", "Arial", sans-serif; }',
			message: messages.rejected('sans-serif'),
			line: 1,
			column: 42,
```

**Review by ybiquitous — COMMENTED** [review]
[suggest] Can you move `fixed:` next to `code:` like other cases?

File: lib/rules/font-family-name-quotes/index.js [hunk 3]
```diff
@@ -102,12 +132,22 @@ const rule = (primary) => {
...
 			if (keywordSets.fontFamilyKeywords.has(family.toLowerCase()) || isSystemFontKeyword(family)) {
 				if (hasQuotes) {
+					if (context.fix) {
+						const openIndex = fontFamilyNode.sourceIndex;
+						const closeIndex = openIndex + rawFamily.length - 1;
+
+						return {
+							type: 'remove',
+							indexes: [openIndex, closeIndex],
+						};
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:148]: [suggest] It seems a bit hard to read to me that `checkFamilyName()` returns an object or `undefined`. 🤔 

For example, can you call a helper function like `removeQuotes()` as below (`addQuotes()` as well)?

```suggestion
						removeQuotes(decl, fontFamilyNode);
						return;
```

**Review by ybiquitous — COMMENTED** [review]
@MrBrN197 I've left some suggestions. I would appreciate them if you could consider them. 😃

**Comment by jeddy3** [comment]
> So, it doesn't seem to be a big problem, even if font-family-name-quotes always fixes CSS with single quotes (or double quotes).

Yes, let's always fix with double quotes. It's what the standard config uses because double quotes are most commonly used in the specification examples.

**Comment by jeddy3** [comment]
@MrBrN197 Thanks for making those changes.

Can you implement @ybiquitous' [two suggestions](https://github.com/stylelint/stylelint/pull/5806#pullrequestreview-844837296) above, and then I think we'll be ready for a final review?

**Comment by MrBrN197** [comment]
@jeddy3 Hey, Thanks, Yes I'm getting on the other changes.

File: lib/rules/font-family-name-quotes/index.js (hunk 3 above)
Comment (MrBrN197) [reply lib/rules/font-family-name-quotes/index.js]: I'm currently working on this. the declaration value can have multiple font-families. If I add quotes or remove quotes from the `decl.value` the indexes in the reset of the following `fontFamilyNode`s will be offset and be incorrect. so I'm currently still working on this in my free time.

File: lib/rules/font-family-name-quotes/index.js [hunk 4]
```diff
@@ -56,8 +56,67 @@ function quotesRequired(family) {
 		.some((word) => /^(?:-?\d|--)/.test(word) || !/^[-\w\u{00A0}-\u{10FFFF}]+$/u.test(word));
 }
 
+/**
+ * @typedef {{
+ * name: string,
+ * hasQuotes: boolean,
+ * sourceIndex: number,
+ * removeQuotes: function,
+ * addQuotes: function
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:65]: [suggest] How about the following refactoring ideas?

- Make indentation for each property
- Avoid `function` for more clarity

```suggestion
 *   name: string,
 *   hasQuotes: boolean,
 *   sourceIndex: number,
 *   removeQuotes: () => void,
 *   addQuotes: () => void,
```

File: lib/rules/font-family-name-quotes/index.js [hunk 5]
```diff
@@ -56,8 +56,67 @@ function quotesRequired(family) {
...
+ * }} MutableNode
+ */
+
+/**
+ *
+ * @param {import('postcss-value-parser').Node[]} fontFamilies
+ * @param {import('postcss').Declaration} decl
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:72]: [suggest] How about adding `@returns` for more readability and type-safety?

```suggestion
 * @param {import('postcss').Declaration} decl
 * @returns {MutableNode[]}
```

File: lib/rules/font-family-name-quotes/index.js [hunk 6]
```diff
@@ -56,8 +56,67 @@ function quotesRequired(family) {
...
+			resetIndexes: (/** @type {number} */ offset) => {
+				mutableNodes.slice(idx + 1).forEach((n) => (n.sourceIndex += offset));
+			},
+			removeQuotes() {
+				if (this.hasQuotes === false) return;
+
+				const openIndex = this.sourceIndex;
+				const closeIndex = openIndex + this.name.length + 2;
+
+				this.hasQuotes = true;
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:94]: [ask] Is `false` valid?

```suggestion
				this.hasQuotes = false;
```

File: lib/rules/font-family-name-quotes/index.js [hunk 7]
```diff
@@ -69,28 +128,26 @@ const rule = (primary) => {
 		}
 
 		root.walkDecls(/^font(-family)?$/i, (decl) => {
-			const fontFamilies = findFontFamily(decl.value);
+			let fontFamiliess = MakeMutableFontFamilies(findFontFamily(decl.value), decl);
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:131]: [suggest]

- `const` is preferred over `let` here
- `fontFamilyNodes` is more readable than `fontFamilies`

```suggestion
			const fontFamilyNodes = MakeMutableFontFamilies(findFontFamily(decl.value), decl);
```

File: lib/rules/font-family-name-quotes/index.js [hunk 8]
```diff
@@ -69,28 +128,26 @@ const rule = (primary) => {
...
+			for (const fontFamilyNode of fontFamiliess) {
+				checkFamilyName(fontFamilyNode, decl);
 			}
 		});
 
 		/**
-		 * @param {string} rawFamily
-		 * @param {import('postcss').Declaration} decl
+		 * @param { MutableNode } fontFamilyNode
+		 * @param { import('postcss').Declaration} decl
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:144]: [suggest] How about removing extra spaces?

```suggestion
		 * @param {MutableNode} fontFamilyNode
		 * @param {import('postcss').Declaration} decl
```

File: lib/rules/font-family-name-quotes/index.js [hunk 9]
```diff
@@ -56,8 +56,67 @@ function quotesRequired(family) {
...
+ * addQuotes: function
+ * }} MutableNode
+ */
+
+/**
+ *
+ * @param {import('postcss-value-parser').Node[]} fontFamilies
+ * @param {import('postcss').Declaration} decl
+ */
+const MakeMutableFontFamilies = (fontFamilies, decl) => {
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/index.js:74]: [suggest] How about a lower-case function name as below?

```suggestion
const makeMutableFontFamilies = (fontFamilies, decl) => {
```

**Review by ybiquitous — COMMENTED** [review]
@MrBrN197 Thank you and sorry for the late response. Great work. 👍🏼 

I've left some easy refactoring suggestions, so please consider them.

File: lib/rules/font-family-name-quotes/index.js (hunk 7 above)
Comment (MrBrN197) [reply lib/rules/font-family-name-quotes/index.js:132]: thanks

**Comment by MrBrN197** [comment]
@ybiquitous Thanks for the feedback. I've made those changes.

**Review by ybiquitous — APPROVED** [review]
@MrBrN197 Great! LGTM 👍🏼

**Comment by ota-meshi** [comment]
I think we need to mark the rule as `(Autofixable)` in the list of rules.
https://github.com/stylelint/stylelint/blob/main/docs/user-guide/rules/list.md

**Review by ybiquitous — APPROVED** [review]

**Review by jeddy3 — APPROVED** [review]
@MrBrN197 Thank you! LGTM.

**Comment by jeddy3** [comment]
Changelog entry:

> - Added: `font-family-name-quotes` autofix ([#5806](https://github.com/stylelint/stylelint/pull/5806)).

**Pull request merged**